crates/sc-consensus-subspace/src/slot_worker.rs

```diff
@@ -558,6 +600,59 @@ where
             public_key.to_raw_vec()
         )))
     }
+
+    /// Builds the proof of time for the block being proposed.
+    async fn build_block_pot(
+        &self,
+        parent_header: &Block::Header,
+        parent_pre_digest: &PreDigest<FarmerPublicKey, FarmerPublicKey>,
+        slot_number: SlotNumber,
+    ) -> Result<Option<PotPreDigest>, PotCreateError<Block>> {
+        let proof_of_time = match &self.proof_of_time {
+            Some(proof_of_time) => proof_of_time.clone(),
+            _ => {
+                // PoT feature disabled.
+                return Ok(None);
+            }
+        };
+        let block_number = *parent_header.number() + One::one();
+
+        // Block 1 does not have proofs
+        if block_number.is_one() {
+            return Ok(Some(PotPreDigest::FirstBlock(slot_number)));
+        }
+
+        // Block 2 onwards.
+        // Get the start slot number for the proofs in the new block.
+        let parent_pot_digest = parent_pre_digest.proof_of_time.as_ref().ok_or_else(|| {
+            // PoT needs to be present in the block if feature is enabled.
+            PotCreateError::ParentMissingPotDigest {
+                parent_block_number: *parent_header.number(),
+                parent_slot_number: parent_pre_digest.slot.into(),
+                slot_number,
+            }
+        })?;
+
+        proof_of_time
+            .get_block_proofs(block_number.into(), slot_number, parent_pot_digest)
```

Comment:
Ditto .. as we plumb randomness from PoT, this would async block if the PoT chain is behind.

I assume the async claim_slot() would be cancelled by substrate if the proofs are not available in time for this slot, so we won't be producing blocks without proofs in them

Reply:
It will not be magically cancelled, you need to return `None` from `fn claim_slot`. I mean block creation will be aborted later if it takes longer than expected, but it is not the same notion I think.

In fact it is possible that for devnet or local chain we might have periods of time where timekeeper is down, does this mean block production will never be able to continue from there?

Reply:
That is what I meant .. if it is in an async wait, it will get aborted.

>In fact it is possible that for devnet or local chain we might have periods of time where timekeeper is down, does this mean block production will never be able to continue from there?

Yes, that is the expected behavior as it is unable to get the randomness to create the challenge

Reply:
Okay, I started discussion around this on the forum: https://forum.subspace.network/t/notion-of-time-slots-with-pot/1600

Reply:
Thanks, saw that/responded .. don't think that can be solved easily by having slot number = wall clock time